source/GameObjects.cpp

```diff
@@ -0,0 +1,458 @@
+#include "DataFile.h"
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "Effect.h"
+#include "Files.h"
+#include "FillShader.h"
+#include "Fleet.h"
+#include "FogShader.h"
+#include "text/FontSet.h"
+#include "Galaxy.h"
+#include "GameEvent.h"
+#include "Government.h"
+#include "Hazard.h"
+#include "ImageSet.h"
+#include "Interface.h"
+#include "LineShader.h"
+#include "MaskManager.h"
+#include "Minable.h"
+#include "Mission.h"
+#include "Music.h"
+#include "News.h"
+#include "Outfit.h"
+#include "OutlineShader.h"
+#include "Person.h"
+#include "Phrase.h"
+#include "Planet.h"
+#include "PlayerInfo.h"
+#include "PointerShader.h"
+#include "Politics.h"
+#include "Random.h"
+#include "RingShader.h"
+#include "Ship.h"
+#include "Sprite.h"
+#include "SpriteQueue.h"
+#include "SpriteSet.h"
+#include "SpriteShader.h"
+#include "StarField.h"
+#include "StartConditions.h"
+#include "System.h"
+#include "Test.h"
+#include "TestData.h"
+
+#include <algorithm>
+#include <iterator>
+#include <cassert>
+#include <map>
+#include <set>
+#include <utility>
+#include <vector>
+
+class Sprite;
+
+using namespace std;
+
+namespace {
+	// TODO (C++14): make these 3 methods generic lambdas visible only to the CheckReferences method.
+	// Log a warning for an "undefined" class object that was never loaded from disk.
+	void Warn(const string &noun, const string &name)
+	{
+		Files::LogError("Warning: " + noun + " \"" + name + "\" is referred to, but not fully defined.");
+	}
+	// Class objects with a deferred definition should still get named when content is loaded.
+	template <class Type>
+	bool NameIfDeferred(const set<string> &deferred, pair<const string, Type> &it)
+	{
+		if(deferred.count(it.first))
+			it.second.SetName(it.first);
+		else
+			return false;
+
+		return true;
+	}
+	// Set the name of an "undefined" class object, so that it can be written to the player's save.
+	template <class Type>
+	void NameAndWarn(const string &noun, pair<const string, Type> &it)
+	{
+		it.second.SetName(it.first);
+		Warn(noun, it.first);
+	}
+}
+
+
+
+future<void> GameObjects::Load(const vector<string> &sources, bool debugMode)
+{
+	progress = 0.;
+
+	return async(launch::async, [this, sources, debugMode]
+		{
+			vector<string> files;
+			for(const string &source : sources)
+			{
+				// Iterate through the paths starting with the last directory given. That
+				// is, things in folders near the start of the path have the ability to
+				// override things in folders later in the path.
+				auto list = Files::RecursiveList(source + "data/");
+				files.reserve(files.size() + list.size());
+				files.insert(files.end(),
+						make_move_iterator(list.begin()),
+						make_move_iterator(list.end()));
+			}
+
+			const int size = static_cast<int>(files.size()) + 1;
+			for(const auto &path : files)
+			{
+				LoadFile(path, debugMode);
+
+				auto val = progress.load(memory_order::memory_order_relaxed);
+				while(!progress.compare_exchange_weak(val, val + 1. / size));
```

Comment:
Atomics are another feature not yet used and thus likely to be confusing to new contributors, making this worthy of an additional comment.

Further, all empty loop bodies should be noticeable, e.g. put a comment in it stating it is intentionally empty.

(Side note: am I correct in thinking the only failures to update would be due to spurious failures?)

Reply:
> (Side note: am I correct in thinking the only failures to update would be due to spurious failures?)

yes.